### ISSUE TITLE
[OpenMP] kmp_gsupport.cpp fails to build with ` error: use of overloaded operator '[]' is ambiguous (with operand types 'SimpleVLA<kmp_depend_info_t>`

### ISSUE THREAD

**Issue by sylvestre** [llvm/llvm-project#73785]
on linux on trunk
```

      |                 ^
/build/source/openmp/runtime/src/kmp_gsupport.cpp:1286:17: note: built-in candidate operator[](const struct kmp_depend_info *, int)
/build/source/openmp/runtime/src/kmp_gsupport.cpp:1286:17: note: built-in candidate operator[](volatile struct kmp_depend_info *, int)
/build/source/openmp/runtime/src/kmp_gsupport.cpp:1286:17: note: built-in candidate operator[](const volatile struct kmp_depend_info *, int)
/build/source/openmp/runtime/src/kmp_gsupport.cpp:1315:17: error: use of overloaded operator '[]' is ambiguous (with operand types 'SimpleVLA<kmp_depend_info_t>' (aka 'SimpleVLA<kmp_depend_info>') and 'kmp_int32' (aka 'int'))
 1315 |         dep_list[i] = gomp_depends.get_kmp_depend(i);
      |         ~~~~~~~~^~
/build/source/openmp/runtime/src/kmp_utils.h:51:12: note: candidate function
   51 |   const T &operator[](std::size_t Idx) const noexcept { return Ptr[Idx]; }
      |            ^
/build/source/openmp/runtime/src/kmp_utils.h:52:6: note: candidate function
   52 |   T &operator[](std::size_t Idx) noexcept { return Ptr[Idx]; }
      |      ^
/build/source/openmp/runtime/src/kmp_gsupport.cpp:1315:17: note: built-in candidate operator[](struct kmp_depend_info *, int)
 1315 |         dep_list[i] = gomp_depends.get_kmp_depend(i);
```

**Comment by sylvestre** [comment]
@shiltian seems that you cause this regression with:
https://github.com/llvm/llvm-project/commit/97e16da450e94c92456fa5a74768ec1b22fe6b63

**Comment by tuliom** [comment]
I'm seeing this only when the target architecture is i386.

**Comment by shiltian** [comment]
I have fixed the issue by using KMP allocator.

**Comment by shiltian** [comment]
@sylvestre Is it also i386 for you? I can't reproduce it locally, and didn't receive any BB email regarding this.

**Comment by tuliom** [comment]
> I have fixed the issue by using KMP allocator ([97e16da](https://github.com/llvm/llvm-project/commit/97e16da450e94c92456fa5a74768ec1b22fe6b63)) caused by [eaab947](https://github.com/llvm/llvm-project/commit/eaab947a8aa39002e8bdaa82be08cbc31e116a11).

@shiltian I'm afraid you're referring to a different issue.
I can reproduce the error reported in this issue with https://github.com/llvm/llvm-project/commit/97e16da450e94c92456fa5a74768ec1b22fe6b63 .

The logs are available here: https://download.copr.fedorainfracloud.org/results/@fedora-llvm-team/llvm-snapshots-incubator-20231129/fedora-rawhide-i386/06703806-libomp/builder-live.log.gz

**Comment by shiltian** [comment]
> > I have fixed the issue by using KMP allocator ([97e16da](https://github.com/llvm/llvm-project/commit/97e16da450e94c92456fa5a74768ec1b22fe6b63)) caused by [eaab947](https://github.com/llvm/llvm-project/commit/eaab947a8aa39002e8bdaa82be08cbc31e116a11).
> 
> 
> 
> @shiltian I'm afraid you're referring to a different issue.
> 
> I can reproduce the error reported in this issue with https://github.com/llvm/llvm-project/commit/97e16da450e94c92456fa5a74768ec1b22fe6b63 .
> 
> 
> 
> The logs are available here: https://download.copr.fedorainfracloud.org/results/@fedora-llvm-team/llvm-snapshots-incubator-20231129/fedora-rawhide-i386/06703806-libomp/builder-live.log.gz

Yeah, I thought it is `operator new []`.

**Comment by sylvestre** [comment]
yes, i386
Could you please revert your patch until you have a fix?
See:
https://llvm-jenkins.debian.net/job/llvm-toolchain-bullseye-binaries/architecture=i386,distribution=bullseye,label=i386/929/consoleFull

**Comment by shiltian** [comment]
I have re-landed the patch (a4d1d5f). Tested on a 32-bit system (not i386 though).
